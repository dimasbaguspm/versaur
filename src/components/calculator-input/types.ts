import type { CalculatorProps } from '@/components/calculator/types'
import type { ModalRootProps } from '@/components/modal/types'

/**
 * Props for CalculatorInput component
 */
export interface CalculatorInputProps
  extends Omit<CalculatorProps, 'value' | 'onChange'> {
  /**
   * The value of the input (number or empty string)
   */
  value: number | ''
  /**
   * Called when the value changes
   */
  onChange: (value: number | '') => void
  /**
   * Label text to display above the input
   */
  label?: React.ReactNode
  /**
   * Helper text to display below the input
   */
  helperText?: React.ReactNode
  /**
   * Error message for invalid state
   */
  error?: React.ReactNode
  /**
   * Placeholder text for the input
   */
  placeholder?: string
  /**
   * If true, disables the input
   */
  disabled?: boolean
  /**
   * Modal props for customizing the modal behavior
   */
  modalProps?: Partial<ModalRootProps>
  /**
   * Optional: ARIA label for dialog
   */
  'aria-label'?: string
  /**
   * Optional: ARIA labelledby for dialog
   */
  'aria-labelledby'?: string
  /**
   * Optional: ARIA describedby for dialog
   */
  'aria-describedby'?: string
}
