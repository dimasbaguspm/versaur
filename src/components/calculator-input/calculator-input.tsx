import { useRef, useState } from 'react'
import type { FC } from 'react'
import { TextInput } from '@/components/text-input'
import type { CalculatorInputProps } from './types'
import { Calculator } from '@/components/calculator'
import { Modal } from '@/components/modal'
import { Icon } from '../icon'
import { CalculatorIcon } from 'lucide-react'

/**
 * Default label for the modal header
 */
const DEFAULT_HEADER = 'Enter Amount'

/**
 * CalculatorInput is an input field with a modal calculator picker
 */
export const CalculatorInput: FC<
  CalculatorInputProps & { headerLabel?: string }
> = ({
  value,
  onChange,
  placeholder,
  disabled,
  modalProps,
  headerLabel,
  label,
  helperText,
  error,
  ...rest
}) => {
  const [open, setOpen] = useState(false)
  const inputRef = useRef<HTMLInputElement>(null)

  // Helper to handle calculator done
  const handleCalculatorDone = (v: string) => {
    // If empty or invalid, treat as ''
    const trimmed = v.trim()
    if (trimmed === '' || isNaN(Number(trimmed))) {
      onChange('')
    } else {
      onChange(Number(trimmed))
    }
    setOpen(false)
  }

  return (
    <>
      <TextInput
        ref={inputRef}
        type='text'
        label={label}
        value={value === '' ? '' : value}
        onClick={!disabled ? () => setOpen(true) : undefined}
        readOnly
        disabled={disabled}
        error={error}
        helperText={helperText}
        rightContent={<Icon as={CalculatorIcon} size='sm' color='ghost' />}
        placeholder={placeholder}
        aria-label={rest['aria-label']}
        aria-labelledby={rest['aria-labelledby']}
        aria-describedby={rest['aria-describedby']}
        className='cursor-pointer bg-white'
      />
      <Modal
        isOpen={open}
        onOpenChange={setOpen}
        disableOverlayClose={!!disabled}
        disableEscClose={!!disabled}
        size='sm'
        aria-label={rest['aria-label']}
        aria-labelledby={rest['aria-labelledby']}
        aria-describedby={rest['aria-describedby']}
        {...modalProps}
      >
        <Modal.Header>{headerLabel ?? DEFAULT_HEADER}</Modal.Header>
        <Modal.Body className='flex flex-col items-center justify-center p-0'>
          <Calculator
            initialValue={typeof value === 'number' ? value : ''}
            onChange={handleCalculatorDone}
            disabled={disabled}
          />
        </Modal.Body>
      </Modal>
    </>
  )
}
