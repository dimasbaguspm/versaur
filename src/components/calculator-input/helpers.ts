import { cva } from '@/utils/variants'

export const calculatorInputVariants = cva(
  'relative w-full flex items-center rounded-md border border-[var(--color-neutral)] bg-white px-3 py-2 text-base text-[var(--color-foreground)] placeholder-[var(--color-ghost)] focus:outline-none focus:ring-2 focus:ring-[var(--color-primary)] transition',
  {
    variants: {
      disabled: {
        true: 'opacity-50 pointer-events-none',
        false: '',
      },
    },
    defaultVariants: {
      disabled: false,
    },
  }
)
