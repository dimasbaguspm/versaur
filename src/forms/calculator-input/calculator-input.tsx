import { useRef, useState } from 'react'
import type { FC } from 'react'
import { TextInput } from '@/forms/text-input'
import type { CalculatorInputProps } from './types'
import { Calculator } from '@/primitive/calculator'
import { Modal } from '@/overlays/modal'
import { Icon } from '@/primitive/icon'
import { CalculatorIcon } from 'lucide-react'
import { Button } from '@/primitive/button'

/**
 * Default label for the modal header
 */
const DEFAULT_HEADER = 'Enter Amount'

/**
 * CalculatorInput is an input field with a modal calculator picker
 */
export const CalculatorInput: FC<
  CalculatorInputProps & { headerLabel?: string }
> = ({
  value,
  onChange,
  placeholder,
  disabled,
  modalProps,
  headerLabel,
  label,
  helperText,
  error,
  ...rest
}) => {
  const [open, setOpen] = useState(false)
  const inputRef = useRef<HTMLInputElement>(null)
  const [calculatorValue, setCalculatorValue] = useState(
    value === '' ? '' : String(value)
  )
  const [lastValidValue, setLastValidValue] = useState(value ?? '')

  // Only update local state, not parent, during calculator input
  const handleCalculatorChange = (v: string) => {
    setCalculatorValue(v)
    // Try to evaluate the value for local lastValidValue
    let result: number | null = null
    try {
      if (/^[\d+\-*/.\s]+$/.test(v)) {
        const evalResult = eval(v)
        if (typeof evalResult === 'number' && !isNaN(evalResult)) {
          result = evalResult
        }
      }
    } catch {
      // ignore invalid expressions
    }
    if (result !== null) {
      setLastValidValue(result)
    }
  }

  // Sync calculatorValue with value when modal opens
  const handleOpen = () => {
    setCalculatorValue(value === '' ? '' : String(value))
    setOpen(true)
  }

  // Only call parent onChange when OK is pressed
  const handleOk = () => {
    const trimmed = calculatorValue.trim()
    let result: number | null = null
    try {
      if (/^[\d+\-*/.\s]+$/.test(trimmed)) {
        const evalResult = eval(trimmed)
        if (typeof evalResult === 'number' && !isNaN(evalResult)) {
          result = evalResult
        }
      }
    } catch {
      // ignore invalid expressions
    }
    if (result !== null) {
      onChange(result)
      setLastValidValue(result)
    } else {
      onChange(lastValidValue)
    }
    setOpen(false)
  }

  const handleClear = () => {
    setCalculatorValue('')
    setLastValidValue('')
    onChange('')
  }

  return (
    <>
      <TextInput
        ref={inputRef}
        type='text'
        label={label}
        value={value === '' ? '' : value}
        onClick={!disabled ? handleOpen : undefined}
        readOnly
        disabled={disabled}
        error={error}
        helperText={helperText}
        rightContent={<Icon as={CalculatorIcon} size='sm' color='ghost' />}
        placeholder={placeholder}
        aria-label={rest['aria-label']}
        aria-labelledby={rest['aria-labelledby']}
        aria-describedby={rest['aria-describedby']}
        className='cursor-pointer bg-white'
      />
      <Modal
        isOpen={open}
        onOpenChange={setOpen}
        disableOverlayClose={!!disabled}
        disableEscClose={!!disabled}
        size='sm'
        aria-label={rest['aria-label']}
        aria-labelledby={rest['aria-labelledby']}
        aria-describedby={rest['aria-describedby']}
        {...modalProps}
      >
        <Modal.Header>{headerLabel ?? DEFAULT_HEADER}</Modal.Header>
        <Modal.Body className='flex flex-col items-center justify-center pb-0'>
          <Calculator
            initialValue={typeof value === 'number' ? value : ''}
            onChange={handleCalculatorChange}
            disabled={disabled}
          />
        </Modal.Body>
        <Modal.Footer className='pb-4 pt-0'>
          <Button
            variant='neutral-ghost'
            onClick={handleClear}
            disabled={disabled}
          >
            Clear
          </Button>
          <Button
            variant='primary-ghost'
            onClick={handleOk}
            disabled={disabled}
          >
            OK
          </Button>
        </Modal.Footer>
      </Modal>
    </>
  )
}
