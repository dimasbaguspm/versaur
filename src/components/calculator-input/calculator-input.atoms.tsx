import { forwardRef } from 'react'
import type { CalculatorInputProps } from './types'
import { calculatorInputVariants } from './helpers'

/**
 * CalculatorInputField is the input field for CalculatorInput
 */
export interface CalculatorInputFieldProps
  extends Omit<CalculatorInputProps, 'onChange'> {
  onChange: (value: number | '') => void
  onClick?: React.MouseEventHandler<HTMLInputElement>
  onFocus?: React.FocusEventHandler<HTMLInputElement>
  readOnly?: boolean
}

export const CalculatorInputField = forwardRef<
  HTMLInputElement,
  CalculatorInputFieldProps
>(function CalculatorInputField(
  { value, onChange, placeholder, disabled, onClick, onFocus, ...props },
  ref
) {
  return (
    <input
      ref={ref}
      type='text'
      inputMode='decimal'
      className={calculatorInputVariants({ disabled })}
      value={value === '' ? '' : value}
      onChange={e => {
        const val = e.target.value
        if (val === '') onChange('')
        else if (/^-?\d*(\.|,)?\d*$/.test(val)) {
          onChange(Number(val.replace(',', '.')))
        }
      }}
      placeholder={placeholder}
      disabled={disabled}
      aria-disabled={disabled}
      onClick={onClick}
      onFocus={onFocus}
      {...props}
    />
  )
})
