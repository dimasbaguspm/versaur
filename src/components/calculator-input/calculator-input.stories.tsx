/**
 * CalculatorInput stories
 *
 * Demonstrates the CalculatorInput component for numeric entry with a modal calculator
 */
import type { Meta, StoryObj } from '@storybook/react'
import { CalculatorInput } from './calculator-input'
import { useState } from 'react'

const meta: Meta<typeof CalculatorInput> = {
  title: 'Form/CalculatorInput',
  component: CalculatorInput,
  tags: ['autodocs'],
}
export default meta

type Story = StoryObj<typeof CalculatorInput>

export const Basic: Story = {
  render: function BasicStory() {
    const [value, setValue] = useState<number | ''>('')
    return (
      <CalculatorInput
        value={value}
        onChange={setValue}
        label='Enter amount'
        helperText='Enter a value using the calculator'
      />
    )
  },
}

export const WithError: Story = {
  render: function WithErrorStory() {
    const [value, setValue] = useState<number | ''>('')
    return (
      <CalculatorInput
        value={value}
        onChange={setValue}
        label='Amount'
        error={value === '' ? 'Amount is required' : undefined}
      />
    )
  },
}
export const Disabled: Story = {
  render: function DisabledStory() {
    const [value, setValue] = useState<number | ''>('')
    return (
      <CalculatorInput
        value={value}
        onChange={setValue}
        label='Disabled'
        disabled
      />
    )
  },
}
