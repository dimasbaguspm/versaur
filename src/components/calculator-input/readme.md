# CalculatorInput

A calculator input field that opens a modal calculator for numeric entry. Follows the same pattern as TimePickerInput, wrapping the Calculator component in a Modal for accessible, mobile-friendly numeric input.

## Usage

```tsx
import { CalculatorInput } from '@/components/calculator-input'

<CalculatorInput value={amount} onChange={setAmount} placeholder="Enter amount" />
```

- Uses the Modal compound pattern for the calculator popup
- Fully type-safe, accessible, and mobile-friendly
- See `TimePickerInput` for a similar pattern
