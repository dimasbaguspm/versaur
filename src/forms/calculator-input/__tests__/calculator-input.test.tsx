import { render, fireEvent, screen } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../calculator-input.stories'
import { CalculatorInput } from '../calculator-input'

describe('CalculatorInput', () => {
  const { Basic, WithError, Disabled } = composeStories(stories)

  it('renders correctly', () => {
    const { asFragment } = render(<Basic />)
    expect(asFragment()).toMatchSnapshot()
  })

  it('shows helper text and label', () => {
    render(<Basic />)
    expect(screen.getByLabelText('Enter amount')).toBeInTheDocument()
    expect(
      screen.getByText('Enter a value using the calculator')
    ).toBeInTheDocument()
  })

  it('shows error when value is empty', () => {
    render(<WithError />)
    expect(screen.getByText('Amount')).toBeInTheDocument()
    expect(screen.getByText('Amount is required')).toBeInTheDocument()
  })

  it('opens modal on click', () => {
    render(<Basic />)
    const input = screen.getByLabelText('Enter amount')
    fireEvent.click(input)
    expect(screen.getByRole('dialog')).toBeInTheDocument()
  })

  it('is disabled when disabled prop is set', () => {
    render(<Disabled />)
    const input = screen.getByLabelText('Disabled')
    expect(input).toBeDisabled()
  })

  it('performs basic addition and saves result on OK', () => {
    const handleChange = vi.fn()
    render(
      <CalculatorInput
        value={''}
        onChange={handleChange}
        label='Enter amount'
        helperText='Enter a value using the calculator'
      />
    )
    const input = screen.getByLabelText('Enter amount')
    fireEvent.click(input)
    // Simulate calculator button presses: 1 + 2
    fireEvent.click(screen.getByText('1'))
    fireEvent.click(screen.getByText('+'))
    fireEvent.click(screen.getByText('2'))
    // Press OK to save
    fireEvent.click(screen.getByText('OK'))
    expect(handleChange).toHaveBeenCalledWith(3)
  })
})
