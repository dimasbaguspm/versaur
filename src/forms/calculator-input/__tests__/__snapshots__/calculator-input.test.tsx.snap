// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CalculatorInput > renders correctly 1`] = `
<DocumentFragment>
  <div>
    <label
      class="block text-sm font-medium text-foreground mb-2"
      for="«r0»"
    >
      Enter amount
    </label>
    <div
      class="relative w-full cursor-pointer bg-white"
    >
      <input
        aria-invalid="false"
        class="block w-full rounded-md border bg-white transition-colors focus:outline-none focus:ring-2 focus:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none disabled:bg-gray-50 border-primary/30 text-foreground focus:border-primary focus:ring-primary/20 pl-3 pr-9 h-10"
        id="«r0»"
        readonly=""
        type="text"
        value=""
      />
      <span
        class="absolute right-2.5 top-0 bottom-0 pointer-events-none text-gray-500 flex items-center justify-center w-5"
        data-testid="right-content"
      >
        <svg
          aria-hidden="true"
          class="lucide lucide-calculator inline-flex items-center justify-center w-4 h-4 text-ghost"
          fill="none"
          height="24"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <rect
            height="20"
            rx="2"
            width="16"
            x="4"
            y="2"
          />
          <line
            x1="8"
            x2="16"
            y1="6"
            y2="6"
          />
          <line
            x1="16"
            x2="16"
            y1="14"
            y2="18"
          />
          <path
            d="M16 10h.01"
          />
          <path
            d="M12 10h.01"
          />
          <path
            d="M8 10h.01"
          />
          <path
            d="M12 14h.01"
          />
          <path
            d="M8 14h.01"
          />
          <path
            d="M12 18h.01"
          />
          <path
            d="M8 18h.01"
          />
        </svg>
      </span>
    </div>
    <div
      class="mt-1 text-sm text-gray-600"
    >
      Enter a value using the calculator
    </div>
  </div>
  <div
    aria-modal="true"
    class="fixed inset-0 z-51 transition-opacity duration-300 bg-black/30 backdrop-blur-md flex items-center justify-center transition-opacity duration-200 ease-in-out items-center transition-opacity duration-200 ease-in-out opacity-0 pointer-events-none"
    placement="center"
    role="presentation"
    tabindex="-1"
  >
    <div
      aria-hidden="true"
      class="bg-white rounded-lg shadow-xl relative flex flex-col outline-none mx-4 transition-all duration-200 ease-in-out  w-[20rem] max-w-sm transition-all duration-200 ease-in-out opacity-0 scale-95 pointer-events-none"
      role="dialog"
      tabindex="0"
    >
      <div
        class="px-6 pt-6 pb-2 text-lg font-semibold"
      >
        Enter Amount
      </div>
      <div
        class="px-6 py-2 flex flex-col items-center justify-center pb-0"
      >
        <div
          aria-label="Calculator"
          class="flex flex-col w-full max-w-xs rounded-lg bg-background"
          role="region"
        >
          <input
            aria-label="Calculator input"
            class="w-full mb-3 px-3 py-2 rounded border border-[var(--color-neutral)] bg-[var(--color-neutral-soft)] text-right text-xl font-mono focus:outline-none focus:ring-2 focus:ring-[var(--color-primary)]"
            inputmode="decimal"
            value=""
          />
          <div
            class="flex flex-col gap-1 mb-3"
          >
            <div
              class="flex w-full"
            >
              <button
                aria-label="7"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                7
              </button>
              <button
                aria-label="8"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                8
              </button>
              <button
                aria-label="9"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                9
              </button>
              <button
                aria-label="/"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
                type="button"
              >
                /
              </button>
            </div>
            <div
              class="flex w-full"
            >
              <button
                aria-label="4"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                4
              </button>
              <button
                aria-label="5"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                5
              </button>
              <button
                aria-label="6"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                6
              </button>
              <button
                aria-label="*"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
                type="button"
              >
                *
              </button>
            </div>
            <div
              class="flex w-full"
            >
              <button
                aria-label="1"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                1
              </button>
              <button
                aria-label="2"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                2
              </button>
              <button
                aria-label="3"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                3
              </button>
              <button
                aria-label="-"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
                type="button"
              >
                -
              </button>
            </div>
            <div
              class="flex w-full"
            >
              <button
                aria-label="0"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                0
              </button>
              <button
                aria-label="."
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-neutral text-foreground hover:bg-neutral-soft hover:text-neutral"
                type="button"
              >
                .
              </button>
              <button
                aria-label="="
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-primary text-white hover:bg-primary-soft hover:text-primary"
                type="button"
              >
                =
              </button>
              <button
                aria-label="+"
                class="flex-1 min-w-0 h-12 m-0.5 rounded text-lg font-medium transition-colors select-none bg-secondary text-white hover:bg-secondary-soft hover:text-secondary"
                type="button"
              >
                +
              </button>
            </div>
          </div>
        </div>
      </div>
      <div
        class="px-6 py-4 flex justify-end gap-2 pb-4 pt-0"
      >
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none text-foreground bg-white hover:bg-neutral/70 focus-visible:ring-foreground-soft focus-visible:ring-offset-white h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          Clear
        </button>
        <button
          aria-disabled="false"
          class="inline-flex items-center justify-center gap-2 rounded-md font-medium transition-all duration-200 focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-offset-2 disabled:opacity-50 disabled:pointer-events-none active:scale-[0.98] select-none text-primary bg-white hover:bg-primary-light focus-visible:ring-primary-light focus-visible:ring-offset-white h-10 px-4 text-sm min-w-[2.5rem]"
          type="button"
        >
          OK
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
